### PR TITLE
Charlesmchen/yap database crash1

### PR DESCRIPTION
**Pull request by charlesmchen**
Comments inline.

PTAL @michaelkirk

### CODE REVIEW

File: Podfile [hunk 1]
```diff
@@ -21,7 +21,7 @@ def shared_pods
   #pod 'JSQMessagesViewController',  git: 'https://github.com/WhisperSystems/JSQMessagesViewController.git', branch: 'signal-master', :inhibit_warnings => true
   #pod 'JSQMessagesViewController',   path: '../JSQMessagesViewController'
   pod 'Mantle', :inhibit_warnings => true
-  pod 'YapDatabase/SQLCipher', '~> 2.9.3', :inhibit_warnings => true
+  pod 'YapDatabase/SQLCipher', :inhibit_warnings => true
```
Comment (charlesmchen) [inline Podfile]: Updates YapDatabase.

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 2]
```diff
@@ -2790,6 +2790,15 @@ - (YapDatabaseConnection *)editingDatabaseConnection
     return _editingDatabaseConnection;
 }
 
+- (void)yapDatabaseModifiedExternally:(NSNotification *)notification
+{
+    OWSAssert([NSThread isMainThread]);
+
+    DDLogVerbose(@"%@ %s", self.logTag, __PRETTY_FUNCTION__);
+
+    [self resetMappings];
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: So, we can't do incremental updates from external modifications, so we just reset our mapping and then reload our view contents.

File: Signal/src/util/MainAppContext.m [hunk 3]
```diff
@@ -12,6 +12,75 @@
...
+                                                 name:UIApplicationWillResignActiveNotification
+                                               object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(applicationDidBecomeActive:)
+                                                 name:UIApplicationDidBecomeActiveNotification
+                                               object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(applicationWillTerminate:)
+                                                 name:UIApplicationWillTerminateNotification
+                                               object:nil];
```
Comment (charlesmchen) [inline Signal/src/util/MainAppContext.m]: I've modified both app contexts to log around their lifecycle.

File: Signal/src/util/Pastelog.m [hunk 4]
```diff
@@ -58,6 +58,14 @@ +(void)submitLogs {
                                          handler:^(UIAlertAction *_Nonnull action) {
                                              [Pastelog.sharedManager sendToSelf:urlString];
                                          }]];
+            [alert addAction:[UIAlertAction
+                                 actionWithTitle:
+                                     NSLocalizedString(@"DEBUG_LOG_ALERT_OPTION_SEND_TO_LAST_THREAD",
+                                         @"Label for the 'send to last thread' option of the the debug log alert.")
+                                           style:UIAlertActionStyleDefault
+                                         handler:^(UIAlertAction *_Nonnull action) {
+                                             [Pastelog.sharedManager sendToLastThread:urlString];
```
Comment (charlesmchen) [inline Signal/src/util/Pastelog.m]: In DEBUG builds, I usually want to send the debug log to myself via a group which is usually the first thread.

File: SignalMessaging/Views/ThreadViewHelper.m [hunk 5]
```diff
@@ -48,49 +52,117 @@ - (void)initializeMapping
...
-        dispatch_async(dispatch_get_main_queue(), ^{
-            [weakSelf updateThreads];
-            [weakSelf.delegate threadListDidChange];
-        });
-    }];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(applicationWillEnterForeground:)
+                                                 name:UIApplicationWillEnterForegroundNotification
+                                               object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self
```
Comment (charlesmchen) [inline SignalMessaging/Views/ThreadViewHelper.m]: Its expensive to rebuild our mapping every time we get an "external database modification" event, so stop observing DB changes in the main app when it is in the background and rebuild our mapping whenever we re-enter the foreground.

Comment (charlesmchen) [reply SignalMessaging/Views/ThreadViewHelper.m]: Note that home and conversation view already do this.

File: Signal/src/util/Pastelog.m [hunk 6]
```diff
@@ -261,6 +269,29 @@ - (void)sendToSelf:(NSString *)url
             }];
         [ThreadUtil sendMessageWithText:url inThread:thread messageSender:messageSender];
     });
+
+    // Also copy to pasteboard.
+    [[UIPasteboard generalPasteboard] setString:url];
+}
+
+- (void)sendToLastThread:(NSString *)url
```
Comment (michaelkirk) [inline Signal/src/util/Pastelog.m]: maybe `sendToMostRecentThread`

File: SignalServiceKit/src/Messages/OWSMessageReceiver.m [hunk 7]
```diff
@@ -13,9 +13,11 @@
 #import "TSStorageManager.h"
 #import "TSYapDatabaseObject.h"
 #import "Threading.h"
+#import <YapDatabase/YapDatabaseAutoView.h>
 #import <YapDatabase/YapDatabaseConnection.h>
 #import <YapDatabase/YapDatabaseTransaction.h>
 #import <YapDatabase/YapDatabaseView.h>
```
Comment (michaelkirk) [inline SignalServiceKit/src/Messages/OWSMessageReceiver.m]: nit: I think the `YapDatabaseView.h` imports can go away from most (all?) files.

File: Signal/src/util/Pastelog.m (hunk 6 above)
Comment (charlesmchen) [reply Signal/src/util/Pastelog.m]: Done.

File: SignalServiceKit/src/Messages/OWSMessageReceiver.m (hunk 7 above)
Comment (charlesmchen) [reply SignalServiceKit/src/Messages/OWSMessageReceiver.m]: Done.

**Pull request merged**